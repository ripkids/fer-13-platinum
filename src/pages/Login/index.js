import { useState } from 'react';
import { useDispatch } from 'react-redux';

import api from '../../api';
import { registerAuth } from '../../store/Auth';

const Login = () => {
  const dispatch = useDispatch();
  const [isError, setIsError] = useState(false);
  const [credential, setCredential] = useState({ email: '', password: '' });
    
  const loginToServer = async () => {
    try {
      setIsError(false);

      const response = await api.loginAdmin(credential);
      const { data } = response;
      
      dispatch(registerAuth(data));
    } catch (error) {
      setIsError(true);
    }
  };
  
  return (
    <>
      <div className="p-5 d-flex flex-column">
        {
          isError && (
            <div
              data-testid="alert-login"
              className="mb-5 p-5 d-flex flex-row rounded-5 bg-danger"
            >
                This is Error
            </div>
          )
        }
        <input
          data-testid="email-input"
          placeholder="Enter email"
          className="mb-5"
          onChange={(ev) =>
            setCredential((prevState) => ({
              ...prevState,
              email: ev.target.value
            }))
          }
        />
        <input
          data-testid="password-input"
          placeholder="Enter password"
          className="mb-5"
          onChange={(ev) =>
            setCredential((prevState) => ({
              ...prevState,
              password: ev.target.value
            }))
          }
        />
        <button
          data-testid="login-button"
          className="btn btn-success"
          onClick={loginToServer}
        >
            Login
        </button>
      </div>
    </>
  );
};

export default Login;