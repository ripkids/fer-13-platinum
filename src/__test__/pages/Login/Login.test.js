import renderer from 'react-test-renderer';
import {
  describe,
  expect,
  test
} from '@jest/globals';
import {
  fireEvent,
  screen,
  waitFor
} from '@testing-library/react';

import {
  rawRenderWithProviders,
  renderWithProviders
} from '../../../mocks/redux';
import { mockLocalStorage } from '../../../mocks/localStorage';
import Login from '../../../pages/Login';

describe('Login', () => {
  const { setItemMock } = mockLocalStorage();

  test('Login component should match with snapshot', () => {
    const component = renderer
      .create(
        rawRenderWithProviders(<Login />)
      )
      .toJSON();

    expect(component).toMatchSnapshot();
  });
  
  test('Alert should be visible when loginAdmin failed',
    async () => {
      renderWithProviders(<Login />);

      const emailInput = screen.getByTestId('email-input');
      const passwordInput = screen.getByTestId('password-input');
      const loginButton = screen.getByTestId('login-button');

      fireEvent.change(
        emailInput,
        {
          target: {
            value: 'test@test.id'
          }
        }
      );
      fireEvent.change(
        passwordInput,
        {
          target: {
            value: '123456'
          }
        }
      );
      fireEvent.click(loginButton);

      await waitFor(() => {
        expect(screen.getByTestId('alert-login')).toBeVisible();
      });
    });

  test('token in localStorage should match with expected value',
    async () => {
      renderWithProviders(<Login />);
      
      const emailInput = screen.getByTestId('email-input');
      const passwordInput = screen.getByTestId('password-input');
      const loginButton = screen.getByTestId('login-button');

      fireEvent.change(
        emailInput,
        {
          target: {
            value: 'admin@bcr.io'
          }
        }
      );
      fireEvent.change(
        passwordInput,
        {
          target: {
            value: '123456'
          }
        }
      );
      fireEvent.click(loginButton);

      await waitFor(() => {
        expect(setItemMock).toHaveBeenCalledWith('token', 'dummy-token');
      });
    });
});