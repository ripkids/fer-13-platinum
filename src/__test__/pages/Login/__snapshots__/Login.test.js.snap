// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Login component should match with snapshot 1`] = `
<div
  className="p-5 d-flex flex-column"
>
  <input
    className="mb-5"
    data-testid="email-input"
    onChange={[Function]}
    placeholder="Enter email"
  />
  <input
    className="mb-5"
    data-testid="password-input"
    onChange={[Function]}
    placeholder="Enter password"
  />
  <button
    className="btn btn-success"
    data-testid="login-button"
    onClick={[Function]}
  >
    Login
  </button>
</div>
`;
